### PR TITLE
ENH: improve error handling

### PR DESCRIPTION
**Pull request by SYNchroACK**

### CODE REVIEW

File: intelmq/lib/bot.py [hunk 1]
```diff
@@ -138,35 +141,46 @@ def start(self, starting=True, error_on_pipeline=True,
                     self.logger.info("Current Message(event): {!r}."
                                      "".format(self.current_message)[:500])
 
-                self.error_retries_counter += 1
+            except KeyboardInterrupt:
+                self.logger.error("Received KeyboardInterrupt.")
+                self.stop(forced=True)
+                break
```
Comment (sebix) [inline intelmq/lib/bot.py]: Unnecessary line. `stop()` calls `terminate()` and terminate does `exit()`.

File: intelmq/lib/bot.py [hunk 2]
```diff
@@ -138,35 +141,46 @@ def start(self, starting=True, error_on_pipeline=True,
...
+
+                    # reached the maximum number of times
+                    else:
                         if self.parameters.error_dump_message:
                             self.dump_message()
+
+                        # remove message from pipeline
                         self.acknowledge_message()
+
+                        # FIXME: depend of config
```
Comment (sebix) [inline intelmq/lib/bot.py]: Please fix this.

**Comment by SYNchroACK** [comment]
@sebix can you help me check why Travis fail? I think its related to lib/tests.py

**Comment by SYNchroACK** [comment]
Thanks you @mauroasilva for clarifying the error handling concept :)

**Comment by SYNchroACK** [comment]
```
$ python3.4 -m intelmq.botparsers.hphosts.parser hphosts-parseror
Traceback (most recent call last):
  File "/usr/lib/python3.4/runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.4/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.4/dist-packages/intelmq-1.0.0-py3.4.egg/intelmq/bots/parsers/hphosts/parser.py", line 55, in <module>
    bot.start()
  File "/usr/local/lib/python3.4/dist-packages/intelmq-1.0.0-py3.4.egg/intelmq/lib/bot.py", line 141, in start
    self.dump_message()
  File "/usr/local/lib/python3.4/dist-packages/intelmq-1.0.0-py3.4.egg/intelmq/lib/bot.py", line 262, in dump_message
    new_dump_data[timestamp]["traceback"] = traceback.format_exc()
  File "/usr/lib/python3.4/traceback.py", line 256, in format_exc
    return "".join(format_exception(*sys.exc_info(), limit=limit, chain=chain))
  File "/usr/lib/python3.4/traceback.py", line 181, in format_exception
    return list(_format_exception_iter(etype, value, tb, limit, chain))
  File "/usr/lib/python3.4/traceback.py", line 146, in _format_exception_iter
    for value, tb in values:
  File "/usr/lib/python3.4/traceback.py", line 125, in _iter_chain
    context = exc.__context__
AttributeError: 'NoneType' object has no attribute '__context__'
```
sel.dump_message() - doest have access to exception variable


Specific line: https://github.com/SYNchroACK/intelmq/blob/error_handling/intelmq/lib/bot.py#L258

**Comment by sebix** [comment]
> sel.dump_message() - doest have access to exception variable

Which exception variable? The [traceback module](https://docs.python.org/3/library/traceback.html) reads the last traceback from `sys.last_traceback`.

**Comment by sebix** [comment]
You can deactivate the message dumping by adding this to `conf` in `mocked_config` in `lib/test.py` and `tests/lib/bot.py`:

    "error_dump_message": False,

but it does not solve the problem. `dump_message` is only called when an error occured and it is catched.

What errors do we see?

> test-bot: INFO: Start processing.
test-bot: DEBUG: Receiving Message.
test-bot: ERROR: Bot has found a problem.
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 96, in start
    self.process()
  File "/home/sebastian/dev/intelmq/intelmq/tests/bots/test_dummy_bot.py", line 40, in process
    report = self.receive_message()
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 239, in receive_message
    message = self.source_pipeline.receive()
  File "/home/sebastian/dev/intelmq/intelmq/lib/pipeline.py", line 198, in receive
    first_msg = self.state[self.source_queue].pop(0)
IndexError: pop from empty list
test-bot: INFO: Last Correct Message(event): None.
test-bot: INFO: Current Message(event): None.

In many cases, the error occurs on the second message, the processing of the first succeeded.

Another maybe related issue I see is the memory consumption. My nosetests processes got killed after consumed huge amounts of memory and no processes in tests. Looks like a memory leak in a infinite loop...

Please don't forget to apply PEP8 to your code changes after fixing the issues.

**Comment by SYNchroACK** [comment]
> You can deactivate the message dumping by adding this to conf in mocked_config in lib/test.py and tests/lib/bot.py:
> 
> "error_dump_message": False,
> but it does not solve the problem. dump_message is only called when an error occured and it is catched.
> 
> What errors do we see?
> 
> test-bot: INFO: Start processing.
> test-bot: DEBUG: Receiving Message.
> test-bot: ERROR: Bot has found a problem.
> Traceback (most recent call last):
> File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 96, in start
> self.process()
> File "/home/sebastian/dev/intelmq/intelmq/tests/bots/test_dummy_bot.py", line 40, in process
> report = self.receive_message()
> File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 239, in receive_message
> message = self.source_pipeline.receive()
> File "/home/sebastian/dev/intelmq/intelmq/lib/pipeline.py", line 198, in receive
> first_msg = self.state[self.source_queue].pop(0)
> IndexError: pop from empty list
> test-bot: INFO: Last Correct Message(event): None.
> test-bot: INFO: Current Message(event): None.
> In many cases, the error occurs on the second message, the processing of the first succeeded.
> 
> Another maybe related issue I see is the memory consumption. My nosetests processes got killed after consumed huge amounts of memory and no processes in tests. Looks like a memory leak in a infinite loop...
> 
> Please don't forget to apply PEP8 to your code changes after fixing the issues.

It seems to me that is an isse in Pythonlist class, right? I think we should merge

**Comment by sebix** [comment]
> It seems to me that is an isse in Pythonlist class, right? I think we should merge

Veto! (if possible)

We introduced testing because we needed reliability and a framework checks if some component brakes other parts without notice. We introduced travis to perform these checks on all pull requests, so master will never have broken tests.

Merging commits that brake nearly all tests, means perverting the introduced techniques along with the efforts of one month.

> It seems to me that is an isse in Pythonlist class, right?

It worked without these changes on the error handling. You changed something. It does not work. If you change on part of the system, you also have to adapt other parts, as they relied on another behavior. There are two possibilities:

- Your code does brake the functionality of the bots.
- Your code does brake the testing components, which need to get adapted to the new behavior.

In both cases, fixes are needed for this PR. There is no reason to merge broken code.

**Comment by sebix** [comment]
Previously, the bot stopped after one `process()` run, if `error_retries_counter` was `>= error_max_retries` and `error_max_retries` as 0. Apparently, this is no longer the case, it is now a `>` check, which is wrong in my opinion.

https://github.com/certtools/intelmq/blob/master/intelmq/lib/test.py#L40
https://github.com/certtools/intelmq/pull/345/files#diff-c700537ee2ab53539af3005dedb87890R138

**Comment by SYNchroACK** [comment]
> Veto! (if possible)

LoL? if possible? ROFL... 

...slows down man... relax

**Comment by SYNchroACK** [comment]
> Previously, the bot stopped after one process() run, if error_retries_counter was >= error_max_retries and error_max_retries as 0. Apparently, this is no longer the case, it is now a > check, which is wrong in my opinion.
> 
> https://github.com/certtools/intelmq/blob/master/intelmq/lib/test.py#L40
> https://github.com/certtools/intelmq/pull/345/files#diff-c700537ee2ab53539af3005dedb87890R138

@sebix hum, im trying to understand why you are saying "is wrong". Can you be more precisely?
https://github.com/synchroack/intelmq/blob/error_handling/intelmq/lib/bot.py

**Comment by sebix** [comment]
The final-cause needs to be changed, and eventually the test.py. You can also change `error_max_retries = -1` in the both bot-tests related file linked above.

Your changes do not honor `exit_on_stop`, which is false for the tests.

Also, you can remove your completely superfluous traceback workaround. Only makes the code blown up.

The loop in start() does not break the loop after the call of `stop()`.

After applying these fixes, the tests will work again.

**Comment by SYNchroACK** [comment]
> The final-cause needs to be changed, and eventually the test.py. You can also change error_max_retries = -1 in the both bot-tests related file linked above.

Still don't understand why.... I already verified multiple times if that 'IF' is wrong and at until now, is completely ok. Since this issue is difficult to explain by text, I present some tables to try to show you my perspective.

The following information presupposes that an exception happened and bot has `error_procedure` = 'stop':

`NE` = number of exceptions
`RC` = self.error_retries_counter
`MR` = self.error_max_retries


NE | RC | MR | Result
--- | --- | --- | ---
1 | 1 | 0 | stop

NE | RC | MR | Result
--- | --- | --- | ---
1 | 1 | 1 | will retry for the first time
2 | 2 | 1 | stop

NE |RC | MR | Result
--- | --- | --- | ---
1 | 1 | 2 | will retry for the first time
2 | 2 | 2 | will retry for the second time
3 | 3 | 2 | stop


> Your changes do not honor exit_on_stop, which is false for the tests.

That option was removed. Check the `error_procedure` parameter:
https://github.com/certtools/intelmq/blob/master/docs/User-Guide.md#conf-defaults-error

> 
> Also, you can remove your completely superfluous traceback workaround. Only makes the code blown up.

Thank you very much for you constructive criticism. Can you suggest something in practice?

> 
> The loop in start() does not break the loop after the call of stop().

Can you specify the line? 
Stop method has `exit(-1)`. Is `break` necessary?
https://github.com/synchroack/intelmq/blob/error_handling/intelmq/lib/bot.py#L202

> 
> After applying these fixes, the tests will work again.

**Comment by sebix** [comment]
> > Your changes do not honor exit_on_stop, which is false for the tests.

> That option was removed. Check the error_procedure parameter
> > The loop in start() does not break the loop after the call of stop().

> Can you specify the line?
> Stop method has exit(-1). Is break necessary?

The exit is a problem. This terminates the process, which should not happen when running tests. Thus I introduced the `exit_on_stop` parameter to prevent the bot from terminating the process.

If stop() does not exit, it will return and we are back in the while-loop in start(). Thus we break.

>> Also, you can remove your completely superfluous traceback workaround. Only makes the code blown up.

> Thank you very much for you constructive criticism. Can you suggest something in practice?

Yes. REMOVE IT. Revert the change.

> The following information presupposes that an exception happened and bot has error_procedure = 'stop':

You need to add this in the 

> Since this issue is difficult to explain by text, I present some tables to try to show you my perspective.

You are assuming that an actual exception happened. But this is **not** the case. Thus, you need error_max_retries = -1 for the tests.

I advise you to work on the actual code instead of doing purely theoretical thoughts.

**Pull request closed without merging**